client/src/app/_services/login.service.ts

```diff
@@ -17,18 +18,55 @@ export class LoginService {
         specialAccount?: SpecialAccount
     };
 
-    isAuthenticated: Boolean;
+    constructor(private http: HttpClient) {
+        setTimeout(() => {
+            if (localStorage.getItem('currentUser')) {
+                // Refresh the token after 50ms to prevent other call.
+                this.refresh().subscribe();
 
-    constructor(private http: HttpClient) { }
+                // Refresh token every 45 minutes
+                Observable.interval(45 * 60 * 1000)
```

Comment:
Why using `Observable.interval` ? Javascript already come with a `setInterval` function.
And we should not use setInterval:
What if the refresh action fail ? 
We should be using only setTimeout:

In `constructor`  (I prefer `ngOnInit`) : check if a user already exists -> If yes: do the refresh action and save a new `setTimeout` in LoginService

In `login` : when the user login and don't fail, save the JWT -> Set the timeout in LoginService (to call refresh)

In `refresh` : If no failure : save the new JWT and restart a timeout. If failure : cancelTimeout and go to login page!

Reply:
Je vais le faire en français:

**Pour les 50ms:**
Si un utilisateur rafraîchit sa page, le constructeur va vérifier si il existe un utilisateur et dans ce cas la, il y en a effectivement un. Il va donc rafraîchir le token. Le problème c'est si la page a besoin de faire des appels au BACK, ils vont utilisé le "vieux" token qui n'a pas eu le temps d'être rafraîchi. Le back renverra alors une erreur 401 car pour lui, le token est révoqué. Il faut donc ajouter un délai pour que les appels de la page passent en premier et le rafraîchissement du token en 2ème.

**Pour `Observable.interval`:**
Aucune raison valable, c'est juste que je trouve ça cool les Observables. Je peux complètement utiliser la fonction setInterval de javascript :)

**Gestion des erreurs**
J'ai effectivement oublié deux trois trucs (notament la redirection vers la page de login), je vais modifier.

Reply:
Ok pour les 50ms, ça change pas grand chose :)